### ISSUE TITLE
Upgrading to Liveblog 1.5 causes PHP Fatal errors

### ISSUE THREAD

**Issue by kevinlisota** [Automattic/liveblog#219]
Just tried upgrading to Liveblog 1.5. It brought down our site with 500 errors.

Specifically, I'm getting this error in our PHP error log: (actual file path obscured with asterisks)

PHP Fatal error:  Call to a member function get() on a non-object in /**/****_/**_/*****/wp-includes/query.php on line 27

I'm having difficulty tracing where liveblog is making this errant call. Any help? We are running WP 4.3.1.

**Comment by philipjohn** [comment]
Do you have a backtrace available at all? That'd help identify where in Liveblog that's originating.

**Comment by kevinlisota** [comment]
I will setup a backtrace on our staging server and report back.




On Sun, Dec 6, 2015 at 12:00 PM -0800, "philipjohn" <notifications@github.com<mailto:notifications@github.com>> wrote:


Do you have a backtrace available at all? That'd help identify where in Liveblog that's originating.

-

**Comment by philipjohn** [comment]
I've just done a quick test on a fresh vanilla install of 4.3.1 and don't get any issues.

Are the errors on the front *and* backend, and for every page?

Perhaps check that the update downloaded properly. You could remove the plugin manually and reinstall to check this.

**Comment by kevinlisota** [comment]
Backtrace didn't actually provide any more details. Simply this message:

Fatal error: Call to a member function get() on a non-object in /nas/content/staging/geekwire/wp-includes/query.php on line 29

I have confirmed that it is not related to our theme, and happens in Twenty Fifteen as well. It is related to at least one plugin, but I haven't isolated it yet. Running without plugins throws now errors.

I can also confirm that the call to get_query_args in query.php is sending the parameter "liveblog" immediately prior to the get() function error on a non-object.

Will report back if I can track this back to a particular plugin.

**Comment by kevinlisota** [comment]
I seemed to have tracked this down to a conflict with WordPress SEO News plugin. (This is the add-on for Yoast SEO news sitemaps)

I'm trying to locate where the conflict is.

**Comment by philipjohn** [comment]
I'd help you debug but it's premium :( Let me know what you find

**Comment by kevinlisota** [comment]
The calling function to get_query_var causing the error is here from a backtrace:

Called by is_entries_ajax_request in WPCOM_Liveblog

**Comment by philipjohn** [comment]
Do you have the full backtrace?

**Comment by kevinlisota** [comment]
Here is a backtrace, generated at the point of failure in query.php.


#0  get_query_var(liveblog) called at [/nas/content/staging/geekwire/wp-content/plugins/liveblog/liveblog.php:415]
#1  WPCOM_Liveblog::is_entries_ajax_request() called at [/nas/content/staging/geekwire/wp-content/plugins/liveblog/liveblog.php:86]
#2  WPCOM_Liveblog::add_custom_post_type_support(WP_Query Object ([query] => Array ([post_type] => any,[post_status] => publish,[meta_query] => Array ([0] => Array ([key] => _yoast_wpseo_newssitemap-standout,[value] => on)),[date_query] => Array ([after] => -7 days)),[query_vars] => Array ([post_type] => any,[post_status] => publish,[meta_query] => Array ([0] => Array ([key] => _yoast_wpseo_newssitemap-standout,[value] => on)),[date_query] => Array ([after] => -7 days),[error] => ,[m] => ,[p] => 0,[post_parent] => ,[subpost] => ,[subpost_id] => ,[attachment] => ,[attachment_id] => 0,[name] => ,[static] => ,[pagename] => ,[page_id] => 0,[second] => ,[minute] => ,[hour] => ,[day] => 0,[monthnum] => 0,[year] => 0,[w] => 0,[category_name] => ,[tag] => ,[cat] => ,[tag_id] => ,[author] => ,[author_name] => ,[feed] => ,[tb] => ,[paged] => 0,[comments_popup] => ,[meta_key] => ,[meta_value] => ,[preview] => ,[s] => ,[sentence] => ,[title] => ,[fields] => ,[menu_order] => ,[category__in] => Array (),[category__not_in] => Array (),[category__and] => Array (),[post__in] => Array (),[post__not_in] => Array (),[post_name__in] => Array (),[tag__in] => Array (),[tag__not_in] => Array (),[tag__and] => Array (),[tag_slug__in] => Array (),[tag_slug__and] => Array (),[post_parent__in] => Array (),[post_parent__not_in] => Array (),[author__in] => Array (),[author__not_in] => Array ()),[tax_query] => WP_Tax_Query Object ([queries] => Array (),[relation] => AND,[] => Array (),[queried_terms] => Array (),[primary_table] => ,[primary_id_column] => ),[meta_query] => ,[date_query] => ,[post_count] => 0,[current_post] => -1,[in_the_loop] => ,[comment_count] => 0,[current_comment] => -1,[found_posts] => 0,[max_num_pages] => 0,[max_num_comment_pages] => 0,[is_single] => ,[is_preview] => ,[is_page] => ,[is_archive] => ,[is_date] => ,[is_year] => ,[is_month] => ,[is_day] => ,[is_time] => ,[is_author] => ,[is_category] => ,[is_tag] => ,[is_tax] => ,[is_search] => ,[is_feed] => ,[is_comment_feed] => ,[is_trackback] => ,[is_home] => 1,[is_404] => ,[is_embed] => ,[is_comments_popup] => ,[is_paged] => ,[is_admin] => ,[is_attachment] => ,[is_singular] => ,[is_robots] => ,[is_posts_page] => ,[is_post_type_archive] => ,[] => ffd359703d04f20d7999dcb8aa2bccd9,[] => ,[thumbnails_cached] => ,[updated_term_meta_cache] => ,[updated_comment_meta_cache] => ,[] => ,[] => Array ([0] => query_vars_hash,[1] => query_vars_changed),[] => Array ([0] => init_query_flags,[1] => parse_tax_query)))
#3  call_user_func_array(Array ([0] => WPCOM_Liveblog,[1] => add_custom_post_type_support), Array ([0] => WP_Query Object ([query] => Array ([post_type] => any,[post_status] => publish,[meta_query] => Array ([0] => Array ([key] => _yoast_wpseo_newssitemap-standout,[value] => on)),[date_query] => Array ([after] => -7 days)),[query_vars] => Array ([post_type] => any,[post_status] => publish,[meta_query] => Array ([0] => Array ([key] => _yoast_wpseo_newssitemap-standout,[value] => on)),[date_query] => Array ([after] => -7 days),[error] => ,[m] => ,[p] => 0,[post_parent] => ,[subpost] => ,[subpost_id] => ,[attachment] => ,[attachment_id] => 0,[name] => ,[static] => ,[pagename] => ,[page_id] => 0,[second] => ,[minute] => ,[hour] => ,[day] => 0,[monthnum] => 0,[year] => 0,[w] => 0,[category_name] => ,[tag] => ,[cat] => ,[tag_id] => ,[author] => ,[author_name] => ,[feed] => ,[tb] => ,[paged] => 0,[comments_popup] => ,[meta_key] => ,[meta_value] => ,[preview] => ,[s] => ,[sentence] => ,[title] => ,[fields] => ,[menu_order] => ,[category__in] => Array (),[category__not_in] => Array (),[category__and] => Array (),[post__in] => Array (),[post__not_in] => Array (),[post_name__in] => Array (),[tag__in] => Array (),[tag__not_in] => Array (),[tag__and] => Array (),[tag_slug__in] => Array (),[tag_slug__and] => Array (),[post_parent__in] => Array (),[post_parent__not_in] => Array (),[author__in] => Array (),[author__not_in] => Array ()),[tax_query] => WP_Tax_Query Object ([queries] => Array (),[relation] => AND,[] => Array (),[queried_terms] => Array (),[primary_table] => ,[primary_id_column] => ),[meta_query] => ,[date_query] => ,[post_count] => 0,[current_post] => -1,[in_the_loop] => ,[comment_count] => 0,[current_comment] => -1,[found_posts] => 0,[max_num_pages] => 0,[max_num_comment_pages] => 0,[is_single] => ,[is_preview] => ,[is_page] => ,[is_archive] => ,[is_date] => ,[is_year] => ,[is_month] => ,[is_day] => ,[is_time] => ,[is_author] => ,[is_category] => ,[is_tag] => ,[is_tax] => ,[is_search] => ,[is_feed] => ,[is_comment_feed] => ,[is_trackback] => ,[is_home] => 1,[is_404] => ,[is_embed] => ,[is_comments_popup] => ,[is_paged] => ,[is_admin] => ,[is_attachment] => ,[is_singular] => ,[is_robots] => ,[is_posts_page] => ,[is_post_type_archive] => ,[] => ffd359703d04f20d7999dcb8aa2bccd9,[] => ,[thumbnails_cached] => ,[updated_term_meta_cache] => ,[updated_comment_meta_cache] => ,[] => ,[] => Array ([0] => query_vars_hash,[1] => query_vars_changed),[] => Array ([0] => init_query_flags,[1] => parse_tax_query)))) called at [/nas/content/staging/geekwire/wp-includes/plugin.php:601]
#4  do_action_ref_array(pre_get_posts, Array ([0] => WP_Query Object ([query] => Array ([post_type] => any,[post_status] => publish,[meta_query] => Array ([0] => Array ([key] => _yoast_wpseo_newssitemap-standout,[value] => on)),[date_query] => Array ([after] => -7 days)),[query_vars] => Array ([post_type] => any,[post_status] => publish,[meta_query] => Array ([0] => Array ([key] => _yoast_wpseo_newssitemap-standout,[value] => on)),[date_query] => Array ([after] => -7 days),[error] => ,[m] => ,[p] => 0,[post_parent] => ,[subpost] => ,[subpost_id] => ,[attachment] => ,[attachment_id] => 0,[name] => ,[static] => ,[pagename] => ,[page_id] => 0,[second] => ,[minute] => ,[hour] => ,[day] => 0,[monthnum] => 0,[year] => 0,[w] => 0,[category_name] => ,[tag] => ,[cat] => ,[tag_id] => ,[author] => ,[author_name] => ,[feed] => ,[tb] => ,[paged] => 0,[comments_popup] => ,[meta_key] => ,[meta_value] => ,[preview] => ,[s] => ,[sentence] => ,[title] => ,[fields] => ,[menu_order] => ,[category__in] => Array (),[category__not_in] => Array (),[category__and] => Array (),[post__in] => Array (),[post__not_in] => Array (),[post_name__in] => Array (),[tag__in] => Array (),[tag__not_in] => Array (),[tag__and] => Array (),[tag_slug__in] => Array (),[tag_slug__and] => Array (),[post_parent__in] => Array (),[post_parent__not_in] => Array (),[author__in] => Array (),[author__not_in] => Array ()),[tax_query] => WP_Tax_Query Object ([queries] => Array (),[relation] => AND,[] => Array (),[queried_terms] => Array (),[primary_table] => ,[primary_id_column] => ),[meta_query] => ,[date_query] => ,[post_count] => 0,[current_post] => -1,[in_the_loop] => ,[comment_count] => 0,[current_comment] => -1,[found_posts] => 0,[max_num_pages] => 0,[max_num_comment_pages] => 0,[is_single] => ,[is_preview] => ,[is_page] => ,[is_archive] => ,[is_date] => ,[is_year] => ,[is_month] => ,[is_day] => ,[is_time] => ,[is_author] => ,[is_category] => ,[is_tag] => ,[is_tax] => ,[is_search] => ,[is_feed] => ,[is_comment_feed] => ,[is_trackback] => ,[is_home] => 1,[is_404] => ,[is_embed] => ,[is_comments_popup] => ,[is_paged] => ,[is_admin] => ,[is_attachment] => ,[is_singular] => ,[is_robots] => ,[is_posts_page] => ,[is_post_type_archive] => ,[] => ffd359703d04f20d7999dcb8aa2bccd9,[] => ,[thumbnails_cached] => ,[updated_term_meta_cache] => ,[updated_comment_meta_cache] => ,[] => ,[] => Array ([0] => query_vars_hash,[1] => query_vars_changed),[] => Array ([0] => init_query_flags,[1] => parse_tax_query)))) called at [/nas/content/staging/geekwire/wp-includes/query.php:2521]
#5  WP_Query->get_posts() called at [/nas/content/staging/geekwire/wp-includes/query.php:4020]
#6  WP_Query->query(Array ([post_type] => any,[post_status] => publish,[meta_query] => Array ([0] => Array ([key] => _yoast_wpseo_newssitemap-standout,[value] => on)),[date_query] => Array ([after] => -7 days))) called at [/nas/content/staging/geekwire/wp-includes/query.php:4131]
#7  WP_Query->__construct(Array ([post_type] => any,[post_status] => publish,[meta_query] => Array ([0] => Array ([key] => _yoast_wpseo_newssitemap-standout,[value] => on)),[date_query] => Array ([after] => -7 days))) called at [/nas/content/staging/geekwire/wp-content/plugins/wpseo-news/classes/class-meta-box.php:190]
#8  WPSEO_News_Meta_Box->standouts_used() called at [/nas/content/staging/geekwire/wp-content/plugins/wpseo-news/classes/class-meta-box.php:202]
#9  WPSEO_News_Meta_Box->standout_description() called at [/nas/content/staging/geekwire/wp-content/plugins/wpseo-news/classes/class-meta-box.php:70]
#10 WPSEO_News_Meta_Box->get_meta_boxes() called at [/nas/content/staging/geekwire/wp-content/plugins/wpseo-news/classes/class-meta-box.php:107]
#11 WPSEO_News_Meta_Box->add_meta_fields_to_wpseo_meta(Array ())
#12 call_user_func_array(Array ([0] => WPSEO_News_Meta_Box Object ([] => Array ([name] => GeekWire,[default_genre] => blog,[default_keywords] => ,[ep_image_src] => https://geekwire.staging.wpengine.com/wp-content/themes/geekwire/assets/img/GeekWire-editors-picks.png,[version] => 3.0,[restrict_sitemap_featured_img] => on,[newssitemap_include_post] => on,[catexclude_contests] => on,[catexclude_geekwork-highlights] => on),[] => 7,[] => ,[] => ),[1] => add_meta_fields_to_wpseo_meta), Array ([0] => Array ())) called at [/nas/content/staging/geekwire/wp-includes/plugin.php:235]
#13 apply_filters(add_extra_wpseo_meta_fields, Array ()) called at [/nas/content/staging/geekwire/wp-content/plugins/wordpress-seo/inc/class-wpseo-meta.php:284]
#14 WPSEO_Meta::init() called at [/nas/content/staging/geekwire/wp-content/plugins/wordpress-seo/wp-seo-main.php:233]
#15 wpseo_init()
#16 call_user_func_array(wpseo_init, Array ([0] => )) called at [/nas/content/staging/geekwire/wp-includes/plugin.php:525]
#17 do_action(plugins_loaded) called at [/nas/content/staging/geekwire/wp-settings.php:276]
#18 require_once(/nas/content/staging/geekwire/wp-settings.php) called at [/nas/content/staging/geekwire/wp-config.php:163]
#19 require_once(/nas/content/staging/geekwire/wp-config.php) called at [/nas/content/staging/geekwire/wp-load.php:37]
#20 require_once(/nas/content/staging/geekwire/wp-load.php) called at [/nas/content/staging/geekwire/wp-blog-header.php:12]
#21 require(/nas/content/staging/geekwire/wp-blog-header.php) called at [/nas/content/staging/geekwire/index.php:17]

**Comment by philipjohn** [comment]
I've followed the code and I *think* I know the cause here.

`get_query_var()` is called by `is_entries_ajax_request()` which is called by `add_custom_post_type_support()`.

`add_custom_post_type_support()` is hooked to `pre_get_posts`. As such, at that time, `$wp_query` has not been set. Therefore, when `get_query_var()` does `global $wp_query` the variable is null. The subsequent attempt to call `$wp_query->get()` fails.

Curiously though, this doesn't appear to be a common issue, which I'd expect from looking at the code. We need to check the circumstances that trigger this fatal more closely.

**Comment by kevinlisota** [comment]
Did you ever make any progress with this? I'm struggling to figure out which plugin is at fault here, whether liveblog or WPSEO News.

I have isolated this a bit further. 

1) wpseo_init is run on the plugins_loaded hook
2) If WPSEO News is present, it runs some meta_box code to insert modifications to the Yoast SEO meta box when you are editing a post in wp-admin
3) During that process, it runs one custom WP_Query called `$standout_query` and then uses found_posts to return a count from that query.
4) This custom WP_Query is running prior to reaching liveblog's add_custom_post_type_support function.
5) When I disable the custom WP_Query, then I do not get the error on get_query_var. It is somehow the result of running a custom WP_Query prior to liveblog's call to `is_entries_ajax_request()`

What I'm unsure of is why this changed in 1.5, and what specifically is causing the failure here. Either one or both plugins may be doing these things on an improper hook order, but I can't seem to figure it out.

**Comment by kevinlisota** [comment]
I figured out why this doesn't happen in the previous version of liveblog.

You changed the order of where liveblog main class is loaded. It used to be loaded in the `plugins_loaded` hook. The conflict reported above does not happen in this instance. 

Implementing the new load order causes the error in `get_query_var` with WPSEO News.

**Comment by sq5gvm** [comment]
I stumbled on this problem today and lost some time figuring this out. Unfortunately found this bug report after locating issue.
In the end I've changed is_entries_ajax_request function to look like this:
`        private static function is_entries_ajax_request() {`
`                global $wp_query;`
`                return !is_null($wp_query) && (bool) get_query_var( self::url_endpoint );`
`        }`

In my case it works ok, but maybe someone can confirm to too.
Maybe this quick and dirty quick fix will be useful for somebody :)

**Comment by philipjohn** [comment]
> 3) During that process, it runs one custom WP_Query called $standout_query and then uses found_posts to return a count from that query.

Ah! I suspect that WP SEO is then not calling `wp_reset_postdata()` leading to a pollution of `$wp_query`

**Comment by kevinlisota** [comment]
Hmm.. That didn't work. I added a `wp_reset_postdata()` after the custom query in WPSEO News, but the error persists.

This seems tied to the load order somehow. In 1.4 of liveblog, you loaded the main class on the `plugins_loaded' hook with a priority of 999. In 1.5, you removed the hook, so the liveblog class is loading earlier, causing earlier calls to `is_entries_ajax_request', which in turn causes the fatal error when SEO News has run its custom query.

**Comment by dfroberg** [comment]
FYI: This is still an issue with Yoast News 3.4 & Liveblog 1.5

**Comment by moorscode** [comment]
@philipjohn I have tracked this problem to the following cause:

Yoast SEO: News is doing a WP_Query on the backend.
This in turn is triggering a `pre_get_posts` which Liveblog is hooked into with `WPCOM_Liveblog::add_custom_post_type_support`. This uses the internal `WPCOM_Liveblog::is_entries_ajax_request` to see if the request is an ajax request.

This calls the `get_query_var` function, which internally uses the global `$wp_query`, which is not accessible on the admin.

So I see two possible fixes:
1. Have an early return in `WPCOM_Liveblog::is_entries_ajax_request` which checks for `is_admin()`.
2. A patch on WordPress core in `get_query_var` to return false when `$wp_query` is not set.

**Comment by kevinlisota** [comment]
@moorscode 
I have attempted the first fix, by changing the `is_entried_ajax_request` function to this:
`private static function is_entries_ajax_request() {
		if ( is_admin() ) {
			return false;
		} else {
			return (bool) get_query_var( self::url_endpoint );
		}
}`

That does fix the fatal errors in wp-admin edit screens. However, it does not fix the same issue occuring on the news sitemap. That still generates a 500 error as follows.

Call to a member function get() on a non-object in /nas/content/live/geekwire/wp-includes/query.php on line 28

**Comment by bfrye26** [comment]
Have you managed to find a solution to this problem. I can not find a good alternative to Yoast news, but the liveblog is so useful for what we do.

**Comment by kevinlisota** [comment]
@bfrye26 I abandoned Yoast SEO and Yoast SEO News & Video. I got tired of their paid products that haven't gotten any meaningful improvements in years, despite the subscription fees. Their core SEO products has become quote bloated as well.

If you use Jetpack, their new Sitemaps capability is quite nice. https://jetpack.com/support/sitemaps/

I had to play around with a few filters to disable the image attachment sitemap and get the correct post types in there for our site, since the UI to control this doesn't really exist yet, but it works great and I finally updated Live Blog.

**Comment by combs** [comment]
We're seeing this manifested as 500s on some redirected pages in WP (when a post is changed from one URL to another). Disabling Yoast SEO News didn't seem to fix the problem; perhaps Yoast itself has something similar these days.

```
2018/10/25 20:29:46 [error] 2044#2044: *438427 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Error: Call to a member function get() on null in WEBROOT/wp-includes/query.php:28
Stack trace:
#0 WEBROOT/html/wp-content/plugins/liveblog/liveblog.php(546): get_query_var('liveblog')
#1 WEBROOT/html/wp-content/plugins/liveblog/liveblog.php(100): WPCOM_Liveblog::is_entries_ajax_request()
#2 WEBROOT/wp-includes/class-wp-hook.php(286): WPCOM_Liveblog::add_custom_post_type_support(Object(WP_Query))
#3 WEBROOT/wp-includes/class-wp-hook.php(310): WP_Hook->apply_filters('', Array)
#4 WEBROOT/wp-includes/plugin.php(515): WP_Hook->do_action(Array)
#5 WEBROOT/wp-includes/class-wp-query.php(1634): do_action_ref_array('pre_get_posts', Array)
#6 WEBROOT/wp-includes/class-wp-query.php(3230): WP_Query->get_posts()
```

If the problem is an unset $wp_query, perhaps liveblog could test for it? We are testing this fix:

``` 
		private static function is_entries_ajax_request() {
			global $wp_query;
			if (isset($wp_query)) {
				return (bool) get_query_var( self::URL_ENDPOINT );
			}
		}
```

**Comment by frenzis** [comment]
I have this issue too in one website.  This error is the same of #604 and #122. The latter show a workaround. 

This issue is opened from a long time. Why don't fix it?

**Comment by philipjohn** [comment]
This same error reared it's head recently in conflict with an entirely different plugin and I took a closer look.

In essence, and as discussed way back in 2014 but [thought to be a core bug](https://github.com/Automattic/liveblog/pull/123#issuecomment-32673846), this error happens whenever another plugin uses `WP_Query` before the main query is created. As @kevinlisota quite [correctly pointed out](https://github.com/Automattic/liveblog/issues/219#issuecomment-198998787) this causes Liveblog's `add_custom_post_type_support()` to be invoked, leading to the `get_query_var()` call that ultimately results in a Fatal Error.

It seemed a little strange to me that `pre_get_posts` was the hook used for adding CPT support so I chose to move it to `after_setup_theme` and in my testing, Liveblog behaved exactly as expected.
